### ISSUE TITLE
Estimator with MirroredStrategy on multiple GPUs seems to start with non-initialized weights on one of the GPUs

### ISSUE THREAD

**Issue by kahkeng** [tensorflow/tensorflow#22364]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: slightly modified example code
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: n/a
- **TensorFlow installed from (source or binary)**: binary (pip)
- **TensorFlow version (use command below)**: ('v1.10.1-0-g4dcfddc5d1', '1.10.1')
- **Python version**: 2.7.12
- **Bazel version (if compiling from source)**: n/a
- **GCC/Compiler version (if compiling from source)**: n/a
- **CUDA/cuDNN version**: 9.0/7.1.4
- **GPU model and memory**: GeForce GTX 1080
- **Exact command to reproduce**:

Slightly modified version of https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/distribute#example-with-estimator-api to add some print nodes.

```

    import tensorflow as tf
    
    def build_model_fn_optimizer():

        def model_fn(features, labels, mode):
            features = tf.Print(features, [features], message="features = ")
            layer = tf.layers.Dense(1, use_bias=True)
            logits = layer(features)
            logits = tf.Print(logits, [logits], message="logits = ")

            if mode == tf.estimator.ModeKeys.PREDICT:
                predictions = {"logits": logits}
                return tf.estimator.EstimatorSpec(mode, predictions=predictions)

            loss = tf.losses.mean_squared_error(
                labels=labels, predictions=tf.reshape(logits, []))

            if mode == tf.estimator.ModeKeys.EVAL:
                return tf.estimator.EstimatorSpec(mode, loss=loss)

            if mode == tf.estimator.ModeKeys.TRAIN:
                train_op = tf.train.GradientDescentOptimizer(0.2).minimize(loss)
                return tf.estimator.EstimatorSpec(mode, loss=loss, train_op=train_op)

        return model_fn
    
    def main(_):
        #distribution = tf.contrib.distribute.OneDeviceStrategy("/device:GPU:0")
        distribution = tf.contrib.distribute.MirroredStrategy(
            ["/device:GPU:0", "/device:GPU:1"]
        )
        config = tf.estimator.RunConfig(train_distribute=distribution)
    
        def input_fn():
            features = tf.data.Dataset.from_tensors([[1.]]).repeat(10)
            labels = tf.data.Dataset.from_tensors(1.).repeat(10)
            return tf.data.Dataset.zip((features, labels))
    
        estimator = tf.estimator.Estimator(
            model_fn=build_model_fn_optimizer(), config=config)
        estimator.train(input_fn=input_fn, steps=10)
    
        eval_result = estimator.evaluate(input_fn=input_fn, steps=10)
        print("Eval result: {}".format(eval_result))
    
        def predict_input_fn():
            predict_features = tf.data.Dataset.from_tensors([[1.]]).repeat(10)
            return predict_features
    
        predictions = estimator.predict(input_fn=predict_input_fn)
        predictions = list(predictions)
        print("Prediction results: {}".format(predictions))
    
    
    if __name__ == "__main__":
        tf.app.run()

```

The output I get is the following:

    WARNING:tensorflow:Using temporary folder as model directory: /tmp/tmpygBKCI
    INFO:tensorflow:Using config: {'_save_checkpoints_secs': 600, '_session_config': None, '_keep_checkpoint_max': 5, '_task_type': 'worker', '_global_id_in_cluster': 0, '_is_chief': True, '_cluster_spec': <tensorflow.python.training.server_lib.ClusterSpec object at 0x7f3bd6649b90>, '_evaluation_master': '', '_save_checkpoints_steps': None, '_keep_checkpoint_every_n_hours': 10000, '_service': None, '_num_ps_replicas': 0, '_tf_random_seed': None, '_master': '', '_device_fn': None, '_num_worker_replicas': 1, '_task_id': 0, '_log_step_count_steps': 100, '_model_dir': '/tmp/tmpygBKCI', '_train_distribute': <tensorflow.contrib.distribute.python.mirrored_strategy.MirroredStrategy object at 0x7f3bd6649b10>, '_save_summary_steps': 100}
    2018-09-18 23:44:42.051140: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
    2018-09-18 23:44:42.244608: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1405] Found device 0 with properties:
    name: GeForce GTX 1080 Ti major: 6 minor: 1 memoryClockRate(GHz): 1.582
    pciBusID: 0000:08:00.0
    totalMemory: 10.92GiB freeMemory: 9.93GiB
    2018-09-18 23:44:42.392404: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:897] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
    2018-09-18 23:44:42.394060: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1405] Found device 1 with properties:
    name: GeForce GTX 1080 Ti major: 6 minor: 1 memoryClockRate(GHz): 1.582
    pciBusID: 0000:42:00.0
    totalMemory: 10.92GiB freeMemory: 10.76GiB
    2018-09-18 23:44:42.398325: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1484] Adding visible gpu devices: 0, 1
    2018-09-18 23:44:42.808254: I tensorflow/core/common_runtime/gpu/gpu_device.cc:965] Device interconnect StreamExecutor with strength 1 edge matrix:
    2018-09-18 23:44:42.808317: I tensorflow/core/common_runtime/gpu/gpu_device.cc:971]      0 1
    2018-09-18 23:44:42.808325: I tensorflow/core/common_runtime/gpu/gpu_device.cc:984] 0:   N Y
    2018-09-18 23:44:42.808331: I tensorflow/core/common_runtime/gpu/gpu_device.cc:984] 1:   Y N
    2018-09-18 23:44:42.808726: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1097] Created TensorFlow device (/device:GPU:0 with 9603 MB memory) -> physical GPU (device: 0, name: GeForce GTX 1080 Ti, pci bus id: 0000:08:00.0, compute capability: 6.1)
    2018-09-18 23:44:42.887416: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1097] Created TensorFlow device (/device:GPU:1 with 10403 MB memory) -> physical GPU (device: 1, name: GeForce GTX 1080 Ti, pci bus id: 0000:42:00.0, compute capability: 6.1)
    INFO:tensorflow:Device is available but not used by distribute strategy: /device:CPU:0
    INFO:tensorflow:Configured nccl all-reduce.
    2018-09-18 23:44:42.999232: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1484] Adding visible gpu devices: 0, 1
    2018-09-18 23:44:42.999351: I tensorflow/core/common_runtime/gpu/gpu_device.cc:965] Device interconnect StreamExecutor with strength 1 edge matrix:
    2018-09-18 23:44:42.999364: I tensorflow/core/common_runtime/gpu/gpu_device.cc:971]      0 1
    2018-09-18 23:44:42.999371: I tensorflow/core/common_runtime/gpu/gpu_device.cc:984] 0:   N Y
    2018-09-18 23:44:42.999379: I tensorflow/core/common_runtime/gpu/gpu_device.cc:984] 1:   Y N
    2018-09-18 23:44:42.999591: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1097] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 9603 MB memory) -> physical GPU (device: 0, name: GeForce GTX 1080 Ti, pci bus id: 0000:08:00.0, compute capability: 6.1)
    2018-09-18 23:44:42.999726: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1097] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:1 with 10403 MB memory) -> physical GPU (device: 1, name: GeForce GTX 1080 Ti, pci bus id: 0000:42:00.0, compute capability: 6.1)
    INFO:tensorflow:Calling model_fn.
    INFO:tensorflow:Calling model_fn.
    INFO:tensorflow:batch_all_reduce invoked for batches size = 2 with algorithm = nccl, num_packs = 1, agg_small_grads_max_bytes = 0 and agg_small_grads_max_group = 10
    INFO:tensorflow:Done calling model_fn.
    INFO:tensorflow:Done calling model_fn.
    INFO:tensorflow:Create CheckpointSaverHook.
    INFO:tensorflow:Graph was finalized.
    2018-09-18 23:44:43.215770: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1484] Adding visible gpu devices: 0, 1
    2018-09-18 23:44:43.215877: I tensorflow/core/common_runtime/gpu/gpu_device.cc:965] Device interconnect StreamExecutor with strength 1 edge matrix:
    2018-09-18 23:44:43.215889: I tensorflow/core/common_runtime/gpu/gpu_device.cc:971]      0 1
    2018-09-18 23:44:43.215897: I tensorflow/core/common_runtime/gpu/gpu_device.cc:984] 0:   N Y
    2018-09-18 23:44:43.215904: I tensorflow/core/common_runtime/gpu/gpu_device.cc:984] 1:   Y N
    2018-09-18 23:44:43.216104: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1097] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 9603 MB memory) -> physical GPU (device: 0, name: GeForce GTX 1080 Ti, pci bus id: 0000:08:00.0, compute capability: 6.1)
    2018-09-18 23:44:43.216251: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1097] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:1 with 10403 MB memory) -> physical GPU (device: 1, name: GeForce GTX 1080 Ti, pci bus id: 0000:42:00.0, compute capability: 6.1)
    INFO:tensorflow:Running local_init_op.
    INFO:tensorflow:Done running local_init_op.
    INFO:tensorflow:Saving checkpoints for 0 into /tmp/tmpygBKCI/model.ckpt.
    features = [[1]]
    logits = [[0]]
    features = [[1]]
    logits = [[-0.225562453]]
    INFO:tensorflow:loss = 1.2510016, step = 0
    features = [[1]]
    logits = [[0]]
    features = [[1]]
    logits = [[0.6646626]]
    features = [[1]]
    features = [[1]]
    logits = [[-5.94375372]]
    logits = [[1.19879758]]
    features = [[1]]
    logits = [[-23.9573383]]
    features = [[1]]
    logits = [[3.89678]]
    features = [[1]]
    features = [[1]]
    logits = [[-82.8739166]]
    logits = [[12.7210035]]
    INFO:tensorflow:Saving checkpoints for 5 into /tmp/tmpygBKCI/model.ckpt.
    INFO:tensorflow:Loss for final step: 3586.108.
    INFO:tensorflow:Calling model_fn.
    INFO:tensorflow:Done calling model_fn.
    INFO:tensorflow:Starting evaluation at 2018-09-18-23:44:45
    INFO:tensorflow:Graph was finalized.
    2018-09-18 23:44:45.322872: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1484] Adding visible gpu devices: 0, 1
    2018-09-18 23:44:45.322972: I tensorflow/core/common_runtime/gpu/gpu_device.cc:965] Device interconnect StreamExecutor with strength 1 edge matrix:
    2018-09-18 23:44:45.322984: I tensorflow/core/common_runtime/gpu/gpu_device.cc:971]      0 1
    2018-09-18 23:44:45.322992: I tensorflow/core/common_runtime/gpu/gpu_device.cc:984] 0:   N Y
    2018-09-18 23:44:45.323000: I tensorflow/core/common_runtime/gpu/gpu_device.cc:984] 1:   Y N
    2018-09-18 23:44:45.323190: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1097] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 9603 MB memory) -> physical GPU (device: 0, name: GeForce GTX 1080 Ti, pci bus id: 0000:08:00.0, compute capability: 6.1)
    2018-09-18 23:44:45.323319: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1097] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:1 with 10403 MB memory) -> physical GPU (device: 1, name: GeForce GTX 1080 Ti, pci bus id: 0000:42:00.0, compute capability: 6.1)
    INFO:tensorflow:Restoring parameters from /tmp/tmpygBKCI/model.ckpt-5
    INFO:tensorflow:Running local_init_op.
    INFO:tensorflow:Done running local_init_op.

Notice that the logits comes out printed as zero the first and third time, which suggests the weights for at least one of the GPU towers is all zeroes at the start (or both are all zeros initially but the prints are showing up interwoven between the two towers).

If I run the code with just a single GPU using OneDeviceStrategy, this does not happen (first logit value printed out is non-zero due to random initial weights).

This affects training for more complicated scenarios since it could lead to incorrect loss computation.

Still happens if I upgrade to TensorFlow 1.11.0-rc1 ('v1.11.0-rc1-0-ge4c4b20805').

I posted on [StackOverflow](https://stackoverflow.com/questions/52396116/tensorflow-estimator-with-mirroredstrategy-on-multiple-gpus-seems-to-start-with) but also decided to open this issue since it seems somewhat similar to the variable initialization issue reported in #19069 which is supposedly fixed. If this is more appropriate on StackOverflow, I can wait for responses there. Thanks!

**Comment by martinwicke** [comment]
Priya, not sure who's best for MirroredStrategy.

**Comment by kahkeng** [comment]
Actually, I tried on 2 other machines (with multiple v100 GPUs and multiple p100 GPUs) and the issue doesn't reproduce there. So this could be something specific to my hardware or software setup for the particular machine above.

**Comment by guptapriya** [comment]
Thanks for the update @kahkeng. Is this something we should investigate further ?

**Comment by kahkeng** [comment]
I think we can close this out, will re-open if this is still an issue, thanks!
